draft-ietf-quic-tls.md

```diff
@@ -255,142 +205,169 @@ A simplified TLS 1.3 handshake with 0-RTT application data is shown in
-of the TLS handshake is carried in Handshake packets.  The Retry packet carries
-a TLS HelloRetryRequest, if it is needed, and Handshake packets carry the
-remainder of the server handshake.
-
-The server sends TLS handshake messages without protection (@H).  The server
-transitions from no protection (@H) to full 1-RTT protection (@1) after it sends
-the last of its handshake messages.
-
-Some TLS handshake messages are protected by the TLS handshake record
-protection.  These keys are not exported from the TLS connection for use in
-QUIC.  QUIC packets from the server are sent in the clear until the final
-transition to 1-RTT keys.
-
-The client transitions from handshake (@H) to 0-RTT keys (@0) when sending 0-RTT
-data, and subsequently to to 1-RTT keys (@1) after its second flight of TLS
-handshake messages.  This creates the potential for unprotected packets to be
-received by a server in close proximity to packets that are protected with 1-RTT
-keys.
-
-More information on key transitions is included in {{hs-protection}}.
+{: #schematic title="QUIC and TLS Interactions"}
 
+Unlike TLS over TCP, QUIC applications which want to send data do not
+send it through TLS "application_data" records. Rather, they send it
+as QUIC STREAM frames which are then carried in QUIC packets.
+
+
+# Carrying TLS Messages {#carrying-tls}
+
+QUIC carries TLS handshake data in CRYPTO_HS frames, each of which
+consists of a contiguous block of handshake data identified by an
+offset and length. Those frames are packaged into QUIC packets
+and encrypted under the current TLS encryption level.
+As with TLS over TCP, once TLS handshake data has
+been delivered to QUIC, it is QUIC's responsibility to deliver it
+reliably. Each chunk of data is associated with the then-current TLS
+sending keys, and if QUIC needs to retransmit that data, it MUST use
+the same keys even if TLS has already updated to newer keys.
+
+One important difference between TLS 1.3 records (used with TCP)
+and QUIC CRYPTO_HS frames is that in QUIC multiple frames may appear
+in the same QUIC packet as long as they are associated with the
+same encryption level. For instance, an implementation might
+bundle a Handshake message and an ACK for some Handshake
+data into the same packet.
+
+In general, the rules for which data can appear in packets of which
+encryption level are the same in QUIC as in TLS over TCP:
+
+- CRYPTO_HS frames MAY appear in packets of any encryption level.
+- CONNECTION_CLOSE and CRYPTO_CLOSE MAY appear in packets of any
+  encryption level other than 0-RTT.
+- PADDING and PING frames MAY appear in packets of any encryption level.
+- ACK frames MAY appear in packets of any encryption level, but
+  MUST only acknowledge packets which appeared in that encryption
+  level.
+- STREAM frames MUST ONLY appear in the 0-RTT and 1-RTT levels.
+- All other frame types MUST only appear at the 1-RTT levels.
```

Comment:
Would it be reasonable to allow the BLOCKED frames in 0-RTT, assuming the previous connection set very small limits in it's TP? For instance, what if the previous connection set the initial maximum number of bidirectional streams to 1, and the client would like to open 2 streams in 0-RTT. Should it be allowed to open 1 stream and include a `STREAM_ID_BLOCKED` frame?

Reply:
I am going to leave this up to MT's discretion.

You should open an issue for BLOCKED.